### PR TITLE
Add `opt_idx` to scheduler config if not assigned by user

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

This is required for #10249 since we need to pass in optimizer_idx in the new hook.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7

### CODE REVIEW

File: pytorch_lightning/core/optimizer.py [hunk 1]
```diff
@@ -337,15 +337,26 @@ def _get_default_scheduler_config() -> Dict[str, Any]:
...
+        "opt_idx": None,  # opt_idx assigned internally if not assigned by user
     }
 
 
-def _validate_scheduler_optimizer(optimizers: List[Any], lr_schedulers: List[Any]) -> None:
-    if any(sch["scheduler"].optimizer not in optimizers for sch in lr_schedulers):
-        raise MisconfigurationException(
-            "Some schedulers are attached with an optimizer that wasn't returned from `configure_optimizers`."
-        )
+def _attach_scheduler_opt_idx(optimizers: List[Any], lr_schedulers: List[Any]) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/core/optimizer.py:344]:
```suggestion
def _set_scheduler_opt_idx(optimizers: List[Any], lr_schedulers: List[Any]) -> None:
```
?

"attach" usually refers to setting a reference to an object or something like that.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/tuner/lr_finder.py [hunk 2]
```diff
@@ -124,10 +128,14 @@ def func(trainer):
...
+            sched_config.update({"scheduler": scheduler, "interval": "step", "opt_idx": 0})
 
-            trainer.strategy.optimizers = [optimizer]
-            trainer.strategy.lr_schedulers = [sched_config]
+            optimizers = [optimizer]
+            lr_schedulers = [sched_config]
+            _attach_scheduler_opt_idx(optimizers, lr_schedulers)
+
+            trainer.strategy.optimizers = optimizers
+            trainer.strategy.lr_schedulers = lr_schedulers
```
Comment (carmocca) [inline pytorch_lightning/tuner/lr_finder.py:138]: Keep the existing structure?

```suggestion
            trainer.strategy.optimizers = [optimizer]
            trainer.strategy.lr_schedulers = [sched_config]
            _attach_scheduler_opt_idx(trainer.strategy.optimizers, trainer.strategy.lr_schedulers)
```

File: pytorch_lightning/core/optimizer.py [hunk 3]
```diff
@@ -337,15 +337,26 @@ def _get_default_scheduler_config() -> Dict[str, Any]:
...
+    for sch in lr_schedulers:
+        opt_found = sch["opt_idx"] is not None
+        if opt_found:
+            continue
+
+        for opt_idx, opt in enumerate(optimizers):
+            if sch["scheduler"].optimizer == opt:
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:351]: Use `is`?

```suggestion
            if sch["scheduler"].optimizer is opt:
```

File: pytorch_lightning/core/optimizer.py [hunk 4]
```diff
@@ -337,15 +337,26 @@ def _get_default_scheduler_config() -> Dict[str, Any]:
         "reduce_on_plateau": False,  # most often not ReduceLROnPlateau scheduler
         "monitor": None,  # value to monitor for ReduceLROnPlateau
         "strict": True,  # enforce that the monitor exists for ReduceLROnPlateau
-        "opt_idx": None,  # necessary to store opt_idx when optimizer frequencies are specified
+        "opt_idx": None,  # opt_idx assigned internally if not assigned by user
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:340]: Should this default to 0 since you removed `Optional` from the annotation in `types.py`?

```suggestion
        "opt_idx": 0,  # opt_idx assigned internally if not assigned by user
```

File: pytorch_lightning/core/optimizer.py [hunk 5]
```diff
@@ -337,15 +337,26 @@ def _get_default_scheduler_config() -> Dict[str, Any]:
...
+
+        for opt_idx, opt in enumerate(optimizers):
+            if sch["scheduler"].optimizer == opt:
+                sch["opt_idx"] = opt_idx
+                opt_found = True
+                break
+
+        if not opt_found:
+            raise MisconfigurationException(
+                "Some schedulers are attached with an optimizer that wasn't returned from `configure_optimizers`."
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:358]: I think this should differentiate between two different errors:

- The user provided an optimizer index but such optimizer does not exist
- The scheduler optimizer reference is not one of the existing optimizers

pseudocode:

```python
for sch in lr_schedulers:
    for opt_idx, opt in enumerate(optimizers):
        if sch["scheduler"].optimizer is opt:
            # assumes we keep the None default
            if sch["opt_idx"] is not None and sch["opt_idx"] != opt_idx:
                error: does not exist
            sch["opt_idx"] = opt_idx
            break
    else:
        error: the reference does not match
```

File: pytorch_lightning/core/optimizer.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/core/optimizer.py:340]: it's required to differentiate from the opt_idx provided by the user in scheduler config and based on your below comment it should stay None

File: pytorch_lightning/core/optimizer.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/core/optimizer.py:358]: ah! good catch

File: pytorch_lightning/core/optimizer.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/core/optimizer.py:340]: I agree. Then you should revert the change to `types.py`

Comment (rohitgr7) [reply pytorch_lightning/core/optimizer.py:340]: just clarifying.. if we update it eventually to an int for all schedulers, won't it be int in `types.py`?? cause we use this config just as an initial entry point.

Comment (carmocca) [reply pytorch_lightning/core/optimizer.py:340]: If the config ever allows None, it should be `Optional[int]`.

There are complex tricks to tell mypy that the type has been narrowed down later (not saying you should do this here): https://www.python.org/dev/peps/pep-0647/

Comment (rohitgr7) [reply pytorch_lightning/core/optimizer.py:340]: thanks :)

File: pytorch_lightning/core/optimizer.py [hunk 6]
```diff
@@ -337,15 +337,27 @@ def _get_default_scheduler_config() -> Dict[str, Any]:
...
-        )
+def _set_scheduler_opt_idx(optimizers: List[Any], lr_schedulers: List[Any]) -> None:
+    for sch in lr_schedulers:
+
+        for opt_idx, opt in enumerate(optimizers):
+            if sch["scheduler"].optimizer is opt:
+                if sch["opt_idx"] is not None and sch["opt_idx"] != opt_idx:
+                    raise MisconfigurationException(
+                        "`opt_idx` set inside scheduler config does not match with index"
+                        " of respective optimizer returned from `configure_optimizers`."
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:352]: Isn't it clearer with "the"s?

```suggestion
                        "`opt_idx` set inside scheduler config does not match with the index"
                        " of the respective optimizer returned from `configure_optimizers`."
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -58,6 +58,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added a `PrecisionPlugin.teardown` method ([#10990](https://github.com/PyTorchLightning/pytorch-lightning/issues/10990))
 
 
+- Added `opt_idx` to scheduler config if not assigned by user ([#11247](https://github.com/PyTorchLightning/pytorch-lightning/issues/11247))
```
Comment (carmocca) [inline CHANGELOG.md:61]: Can you add "Changed" entries for the new misconfiguration checks?

**Review by carmocca — APPROVED** [review]
Cool!

File: CHANGELOG.md (hunk 7 above)
Comment (rohitgr7) [inline CHANGELOG.md:61]:
```suggestion
- Added `opt_idx` to scheduler config if not assigned by user ([#11247](https://github.com/PyTorchLightning/pytorch-lightning/issues/11247))


- Added a `MisconfigurationException` if user provided `opt_idx` in scheduler config doesn't match with actual optimizer index of its respective optimizer ([#11247](https://github.com/PyTorchLightning/pytorch-lightning/issues/11247))
```

File: tests/trainer/optimization/test_optimizers.py [hunk 8]
```diff
@@ -511,6 +511,24 @@ def configure_optimizers(self):
...
+        def configure_optimizers(self):
+            opt1 = optim.SGD(self.layer.parameters(), lr=0.1)
+            opt2 = optim.SGD(self.layer.parameters(), lr=0.1)
+            lr_scheduler = {"scheduler": optim.lr_scheduler.StepLR(opt2, step_size=1), "opt_idx": 0}
+            return [opt1, opt2], [lr_scheduler]
+
+    model = InvalidOptimizerModel()
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True)
+    with pytest.raises(
+        MisconfigurationException, match="`opt_idx` .* does not match with index of respective optimizer"
```
Comment (rohitgr7) [inline tests/trainer/optimization/test_optimizers.py:527]:
```suggestion
        MisconfigurationException, match="`opt_idx` .* does not match with the index of the respective optimizer"
```

File: CHANGELOG.md (hunk 7 above)
Comment (rohitgr7) [reply CHANGELOG.md:61]: done!

**Review by akihironitta — APPROVED** [review]
